lib/Controller/VideoController.php

```diff
@@ -0,0 +1,106 @@
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program. If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+namespace OCA\Viewer\Controller;
+
+use Exception;
+
+use OCP\IRequest;
+use OCP\AppFramework\Http;
+use OCP\AppFramework\Http\DataResponse;
+use OCP\AppFramework\Controller;
+use OCP\L10N\IFactory;
+
+use OCP\Files\File;
+use OCP\Files\IRootFolder;
+use OCP\IUserSession;
+
+class VideoController extends Controller {
+
+	// I copied the annotation below without understanding it... is it useful?
+	/** @var IRootFolder */
+	private $storage;
+
+	// I copied the annotation below without understanding it... is it useful?
+	/** @var array */
+	private $locales;
+
+	public function __construct(
+		string $AppName,
+		IRequest $request,
+		IRootFolder $storage,
+		IFactory $factory,
+		IUserSession $userSession
+	){
+		parent::__construct($AppName, $request);
+		$this->storage = $storage;
+		$this->locales = $factory->findAvailableLocales();;
+		$this->currentUser = $userSession->getUser();
+	}
+
+	// I copied the annotation below without understanding it... is it useful?
+	/**
+	 * @NoAdminRequired
+	 */
+	public function getTracks($videoPath): DataResponse  {
+		// $videoPath is the path to the main video file, eg: 'mypath/movie.mkv'
+		// Return an array of found tracks associated with the main video.
+		// Each track is an object:
+		//   - basename  /* basename of the track */
+		//   - language  /* language for the track */
+		//   - locale    /* (usuallu) 2-letter code for the language */
+		//
+		$video = $this->storage->getUserFolder($this->currentUser->getUID())
+			->get($videoPath);
+		$rootName = pathinfo($video->getFileInfo()->getPath())['filename'];
+		$rootLen = strlen($rootName);
+		// eg: $rootName is 'movie', $rooLen is 5
+		$files = $video->getParent()->getDirectoryListing();
+		// $files if the array of files within the same directory
+		$res = [];
+		foreach ($files as $file) {
+			$path = $file->getFileInfo()->getPath();
+			// eg: $path is 'mypath/.movie.en.vtt'
+			$baseName = pathinfo($path)['basename'];
+			$rootPos = strpos($baseName, $rootName);
+			// eg: $basename is '.movie.en.vtt', $rootPos is 1
+			if (rootPos == 0 || (rootPos == 1 && $baseName[0] == '.')) {
+				$extension = substr($baseName, $rootPos + $rootLen);
+				// eg: $extension is '.en.vtt'
+				$pattern = "/^[.]([^.]+)[.]vtt$/";
+				if (preg_match($pattern, $extension)) {
+					$locale = preg_replace($pattern, "$1", $extension);
+					$key = array_search($locale, array_column($this->locales, 'code'));
+					$language = $key ? $this->locales[$key]['name'] : $locale;
+					array_push($res, [
+						basename => $baseName,
+						language => $language, // eg: 'English'
+						locale => $locale // eg: 'en'
+					]);
+				}
+			}
+		}
```

Comment:
So, I discussed it a bit.
For performances it would be safer to do a [`nodeExists`](https://nextcloud-server.netlify.app/classes/ocp-files-folder#method_nodeExists) for each one of the locale.

Do an array_filter over the list of locale code and that's it.
Then return an array_map to format your results with basename, language and locale so it's safe.

If the folder have 10k files in it, iterating over all of those would be a nightmare :grin:

Reply:
Yes indeed.
And it would have a positive side effect : in case the caption files are readable but the folder itself is not (beardhatcode's scenario), it would enable the player to find the caption files anyway.
On a negative note: it will remove the possibility of "free format" names such as 'movie.en-1.vtt' or 'movie.en-forced.vtt'.

PS -- Yet another idea: a '.movie.aux' folder where the user could store caption files, posters, etc. This folder (if it exists) would be scanned exhaustively. This would be in addition to the scheme you propose.

Reply:
@skjnldsv, any opinion on the above proposal?

Reply:
I don't think your positive side effect will not work, if the folder is not shared those files will not be there. And they shouldn't if the captions are not shared, by the user they should not show up. It must be extremely clear to the user what is being shared.

The negative side effect is a good point, but it might be best to wait for user feedback to see if `movie.en-yolo.vtt` is something users want. I think simply the addition of the captions it such a benefit that users will who want to use them will search for "how to subtitle Nextcloud" and find that they should name their file in a certain way. 

With regard of the `.movie.aux` folder: maybe it is best to first just do it with the filename.vtt for now. Keep it simple, and then once this is merged and running we can see if there is feedback that maybe shines light on another more requested approach. But we should wait for this before making the code more complicated.

Reply:
OK. Let's keep it simple.

Reply:
> With regard of the `.movie.aux` folder: maybe it is best to first just do it with the filename.vtt for now. Keep it simple, and then once this is merged and running we can see if there is feedback that maybe shines light on another more requested approach. But we should wait for this before making the code more complicated.

Also, this is how most media manager works, like plex, sonarr... etc
`filename.langcode.ext` is the standard